### ISSUE TITLE
mempool v1: recovering from broadcasting mempool loop

### ISSUE THREAD

**Issue by alexanderbez** [tendermint/tendermint#6660]
```
ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
```

ref: https://gist.github.com/cmwaters/57a3ae945b58b55c7952a9ee02202c16

1. We should add tracing to the panic recovery.
2. Debug and fix the above issue.

**Comment by cmwaters** [comment]
I saw this error again in an e2e test:
```
validator01    | 2021-07-09T12:09:10Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool stack="goroutine 2384 [running]:\nruntime/debug.Stack(0xc001160bd0, 0x12bac40, 0x1e5db20)\n\truntime/debug/stack.go:24 +0x9f\ngithub.com/tendermint/tendermint/internal/mempool/v1.(*Reactor).broadcastTxRoutine.func1(0xc0001680a0, 0xc001364d80, 0x28)\n\tgithub.com/tendermint/tendermint/internal/mempool/v1/reactor.go:323 +0xf0\npanic(0x12bac40, 0x1e5db20)\n\truntime/panic.go:965 +0x1b9\ngithub.com/tendermint/tendermint/internal/libs/clist.(*CElement).NextWaitChan(0x0, 0x0)\n\tgithub.com/tendermint/tendermint/internal/libs/clist/clist.go:109 +0x41\ngithub.com/tendermint/tendermint/internal/mempool/v1.(*Reactor).broadcastTxRoutine(0xc0001680a0, 0xc001364d80, 0x28, 0xc0016f81f8)\n\tgithub.com/tendermint/tendermint/internal/mempool/v1/reactor.go:383 +0x1f1\ncreated by github.com/tendermint/tendermint/internal/mempool/v1.(*Reactor).processPeerUpdate\n\tgithub.com/tendermint/tendermint/internal/mempool/v1/reactor.go:271 +0x371\n" version=v1
```

**Comment by cmwaters** [comment]
I saw this another time. It seems to actually be coming from the clist when `NextWaitChan()` is called. But the actual line is where the mutex gets called which is weird. How does that become nil?

**Comment by alexanderbez** [comment]
Thanks for sharing @cmwaters -- looking into this now.

**Comment by alexanderbez** [comment]
So there is no way the mutex would be nil -- this stacktace tells me the actual `*CElement` itself is nil, which should not happen.

**Comment by alexanderbez** [comment]
I don't see how this is possible. My hunch tells me maybe https://github.com/tendermint/tendermint/pull/6626 is related. That PR was merged _after_ we first discovered this issue so that leads me to believe it could very well be the culprit.

/cc @JayT106

**Comment by williambanfield** [comment]
Worth noting that in the logs linked in the original comment this happens multiple times in a row:
```
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
[32mvalidator01    |[0m 2021-07-06T12:22:34Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool version=v1
```

**Comment by tessr** [comment]
@williambanfield - you want to do the reversion and then see if it improves?

**Comment by alexanderbez** [comment]
We have two options:

1. Change default or explicitly use `v0` and see if the issue happens again. If so, we know it's 100% #6626 for sure.
2. Revert #6626 and see if the issue happens again. Also gives us a lot of confident, but we can't say 100% for sure it's #6626.

**Comment by JayT106** [comment]
> I don't see how this is possible. My hunch tells me maybe #6626 is related. That PR was merged _after_ we first discovered this issue so that leads me to believe it could very well be the culprit.
> 
> Maybe we revert #6626?
> 
> /cc @JayT106

looks like it had been reverted.  But I agree the issue may relate to my PR. I will fix it if the PR causes the nil pointer to happen.

**Comment by alexanderbez** [comment]
Thank you @JayT106 🙏

**Comment by alexanderbez** [comment]
Closing this issue since we reverted #6626. If we see this again, we can re-open.

**Comment by cmwaters** [comment]
I saw this again in the e2e test last night:
```
validator01    | 2021-07-22T02:37:36Z ERROR recovering from broadcasting mempool loop err="runtime error: invalid memory address or nil pointer dereference" module=mempool stack="goroutine 42477 [running]:\nruntime/debug.Stack(0xc00056abd0, 0x12a2400, 0x1e2db30)\n\truntime/debug/stack.go:24 +0x9f\ngithub.com/tendermint/tendermint/internal/mempool/v1.(*Reactor).broadcastTxRoutine.func1(0xc000c3a960, 0xc00263e378, 0x28)\n\tgithub.com/tendermint/tendermint/internal/mempool/v1/reactor.go:323 +0xf0\npanic(0x12a2400, 0x1e2db30)\n\truntime/panic.go:965 +0x1b9\ngithub.com/tendermint/tendermint/internal/libs/clist.(*CElement).NextWaitChan(0x0, 0x0)\n\tgithub.com/tendermint/tendermint/internal/libs/clist/clist.go:109 +0x41\ngithub.com/tendermint/tendermint/internal/mempool/v1.(*Reactor).broadcastTxRoutine(0xc000c3a960, 0xc00263e378, 0x28, 0xc0027e35f0)\n\tgithub.com/tendermint/tendermint/internal/mempool/v1/reactor.go:383 +0x1f1\ncreated by github.com/tendermint/tendermint/internal/mempool/v1.(*Reactor).processPeerUpdate\n\tgithub.com/tendermint/tendermint/internal/mempool/v1/reactor.go:271 +0x371\n" version=v1
```
Pretty much the same error AFAICS
Here's a link to the test (although I think this expires after a while):
https://pipelines.actions.githubusercontent.com/KyLXqUW5FXEommX1jskveX9BQo621QMUbAS3HHgRXLqddPVNkV/_apis/pipelines/1/runs/43176/signedlogcontent/5?urlExpires=2021-07-22T12%3A52%3A29.4476355Z&urlSigningMethod=HMACV1&urlSignature=26%2Fia6NFaL%2FcmOL3XOFC71dn7UR1OCKE2PE%2FcbIdCbY%3D

**Comment by alexanderbez** [comment]
![image](https://user-images.githubusercontent.com/5265989/126648163-1ec267ee-86c4-4de3-838e-566a1f78ec86.png)


Can we enable `v0` in addition to `v1` and see if we catch the same issue on `v0`? If not, then we'll have to painfully grok the `v1` concurrency :-/

**Comment by alexanderbez** [comment]
Ok so I think the issue is following, and this only pertains to `v1`:


In `v1`, we have an extra level of abstraction, `*WrappedTx`, which wraps a raw tx amongst other very useful things. The mempool data structure uses this abstraction type as the primary type it stores in various indexes.

In the mempool data structure we call `removeTx` which does the following: https://github.com/tendermint/tendermint/blob/master/internal/mempool/v1/mempool.go#L767-L770. 

Concurrently, we also have the reactor gossiping txs via `broadcastTxRoutine`. My guess is that there is a race between https://github.com/tendermint/tendermint/blob/master/internal/mempool/v1/reactor.go#L383 and https://github.com/tendermint/tendermint/blob/master/internal/mempool/v1/mempool.go#L767-L770.

If my theory is correct, then the race happens when trying to broadcast and remove a tx. Since we're removing the tx, we really have no business broadcasting it so i don't think this problem is severe. However, it should still be fixed eventually.

I think we can do one of the following to fix this:

1. Investigate if we can just store the raw tx itself in the `gossipIndex` instead of the `*WrappedTx` (this would make `broadcastTxRoutine` identical to `v0`).
2. Look into locking-based fixes :-/

**Comment by williambanfield** [comment]
@alexanderbez If it's a race between the `removeTx` call and the broadcast routine, I think we could write a test that reproduces this behavior. It would be nondeterministic, but if we ran it enough times we should see the panic. The test would attempt to run this piece of the broadcast routine: https://github.com/tendermint/tendermint/blob/master/internal/mempool/v1/reactor.go#L329-L404 at the same time as the `removeTx` code to try to force a panic. 

It may be worth factoring out the the `broadcastTxRoutine` code inside the for loop into a `fetchNextTx` helper function to help with testing. Starting with this test could help us catch the issue and/or feel more confidence once we've fixed it. Thoughts?

**Comment by alexanderbez** [comment]
It's worth a shot!

I also pushed https://github.com/tendermint/tendermint/pull/6771 in hopes that it might help. It's pretty much very similar to v0 now.

**Comment by williambanfield** [comment]
Ran the test I added in #6772 with the race detector. Before the change that @alexanderbez made in #6771, the test reported a race condition. On the latest master with the change made, the test no longer reports a race. Snippet of test run with some logs removed for brevity:

```
go test . -test.v -run TestReactorBroadcastDoesNotPanic -cpu=3 -race
=== RUN   TestReactorBroadcastDoesNotPanic
2021-08-02T11:26:40-04:00 INFO starting service impl=router service=router
2021-08-02T11:26:40-04:00 INFO starting router channels= listen_addr=0.0.0.0:0 net_addr={"id":"c0c9a1855c202c3fee77c44d1e3c1663122034b6","ip":"0.0.0.0","port":0} node_id=c0c9a1855c202c3fee77c44d1e3c1663122034b6
...
2021-08-02T11:26:40-04:00 INFO closed transport local=c0c9a1855c202c3fee77c44d1e3c1663122034b6
2021-08-02T11:26:40-04:00 DEBUG stopping evict routine
--- PASS: TestReactorBroadcastDoesNotPanic (0.02s)
...
ok      github.com/tendermint/tendermint/internal/mempool/v1    0.443s
```

**Comment by JayT106** [comment]
@cmwaters @williambanfield circle back to check the issue, so my PR might not be the root cause of the panic?
If it's not, maybe can merge it back?
